### ISSUE TITLE
directive stops working to select more files than allowed

### ISSUE THREAD

**Issue by lkdnvc** [danialfarid/ng-file-upload#1388]
ngf-max-files='2' ngf-multiple="true" ngf-keep="true".

1) select 3 file
2) uploader not work (You can not select a file 1 or 2)

[jsfiddle example](http://jsfiddle.net/s8kc7wg0/541/)

this is a mistake, or i can change the behavior?

**Comment by danialfarid** [comment]
You have ngf-max-files = "2" so if you select 3 files they will be all invalid and will be in ngf-model-invalid.

**Comment by danialfarid** [comment]
I will try to make only the extra files invalid in the next version.

**Comment by lkdnvc** [comment]
> You have ngf-max-files = "2" so if you select 3 files they will be all invalid and will be in ngf-model-invalid.

and how can I continue upload?

**Comment by lkdnvc** [comment]
If you set ngf-max-files = "2" and upload 3 files, uploader does not work.

**Comment by ray-peters** [comment]
Ran into the same problem.

**Steps to recreate:**

1. Create a `<ngf-select />` directive with at least the `ngf-multiple="true" ngf-max-files="2" and ngf-keep='true'` attributes.
2. Upload more than 2 files on first attempt, triggering validation.
3. Try to use the `<ngf-select />` directive again.

Looks like the `ngf-keep` attribute is collecting the uploads that are invalid during updateModel() (including uploads that are "invalid" because of maxLength) inside of the `prevInvalidFiles` array. This then appears to breaks subsequent attempts to upload for the following reasons:

- The maxLength validation checks against a var `allLength`, which includes `prevInvalidFiles.length` in the sum.
- The `removeDuplicates()` vets `prevInvalidFiles` from `allNewFiles`.

In conclusion, it appears that it would be best to exclude the maxLength invalids inside the `prevInvalidFiles` array. Replacing `separateInvalids()` with [this gist](https://gist.github.com/ray-peters/565ded3338e8e399cda8) resolves this issue.

**Comment by danialfarid** [comment]
@ray-peters  I don't see how the extra directive will be helpful here.
I would go with this https://github.com/danialfarid/ng-file-upload/issues/1388#issuecomment-189700147 approach

**Comment by ray-peters** [comment]
Hey @danialfarid, thanks for the quick reply.

What should be the expected behavior here? In my opinion, an upload canceled due to reaching the `max-files` limit should not mark the files as invalid.

From my research, the invalid files are tracked in a way that `ngf-keep` will not allow you to upload those files again. Either:

1. The next `max-files` validation check will return false because `max-files` is compared to ` allFiles.length` when it seems `allValidFiles.length` would be more appropriate, and/or
2. If `ngf-keep-distinct === 'true'` then `removeDuplicates()` will vet all `prevInvalidFiles` from any future attempt to upload.

What do you think?

**Comment by danialfarid** [comment]
Let's say max files is 10 and you selected 12 picture, the last two pictures will be invalid and have error and the form will have validation error max files. Since keep is set adding more files would add them to invalid model until you remove some of the files from the model.

**Comment by ray-peters** [comment]
Both ways would work... however I think you might still run into the issues outlined above even if the valids and invalids are partitioned correctly.

**Comment by danialfarid** [comment]
What issues?

**Comment by ahmed-abdulmoniem** [comment]
Guys, any fix on that .. this is not enhancement this is a bug because the control stops working totally if I selected more than the max file limit! It will not work until you refresh the page again!

**Comment by lkdnvc** [comment]
+1

**Comment by ahmed-abdulmoniem** [comment]
+1

**Comment by danialfarid** [comment]
Fixed at 12.0.3
